### PR TITLE
Fixing #404, which prevented connection reuse.

### PR DESCRIPTION
**Pull request by marstr**

### CODE REVIEW

File: storage/client.go [hunk 1]
```diff
@@ -424,9 +424,9 @@ func (c Client) execInternalJSON(verb, url string, headers map[string]string, bo
 	return respToRet, nil
 }
 
-func readResponseBody(resp *http.Response) ([]byte, error) {
-	defer resp.Body.Close()
-	out, err := ioutil.ReadAll(resp.Body)
+func readBody(body io.ReadCloser) ([]byte, error) {
```
Comment (mcardosos) [inline storage/client.go]: I would rename to `readAndCloseBody` to make it more clear. Maybe add a link to the godoc in which it is explained why this must be like that.
Other that that LGTM.

**Comment by jhendrixMSFT** [comment]
Does any of our generated code have this bug?

**Comment by mcardosos** [comment]
Looks like yes :I

**Comment by marstr** [comment]
I'll investigate what it'll take to fix the generated stuff before I close #404. Though depending on the fix, it may be that it makes more sense to file a bug on the generator or in Azure/go-autorest instead of tying this issue.

**Comment by jhendrixMSFT** [comment]
Agreed fixing the generator should be tracked with a separate issue.

**Comment by mcardosos** [comment]
I think we'll need to add a `ByReading()` RespondDecorator. I think it is not needed in funcs that use `ByUnmarshallingJSON()`.

**Comment by marstr** [comment]
Agreed, I took particular care to never apply the `readBody()` (now `readAndCloseBody()`) to any function that has its own use for the response body.

**Comment by marstr** [comment]
I just tinkered around a little (and copied my experiment [here.](https://play.golang.org/p/jnFf7RhqCu)) I verified that there is no harm in reading something that has already been read to completion, but hasn't yet been closed. For that reason, I believe the correct fix for other parts of the repository is to update the `ByClosing()` decorator to first read and discard the contents of `response.Body`.

**Comment by marstr** [comment]
Once I actually started playing around, I decided it may fit the pattern better to create a new decorator called `ByDiscardingBody()`. Thoughts, @mcardosos ?

**Comment by mcardosos** [comment]
I like it. I would prefer not to modify `ByClosing()`.

**Pull request merged**